### PR TITLE
adding nicer help when missing arguments

### PR DESCRIPTION
**Pull request by jasinner**
Signed-off-by: Jason Shepherd jason@jasonshepherd.net

### CODE REVIEW

**Comment by duglin** [comment]
@jasinner can you add some text explaining what you did and why?  For example, is this just going to show the usage stuff w/o the full options list?  And why did you change the exit code from non-zero to zero? Giving people some rationale behind your PR will help move it along.

I agree that the current stuff is lacking and I'm not sure why the more verbose stuff was removed.
For example:
```
$ docker build
docker: "build" requires 1 argument. See '/root/docker/bundles/1.5.0-dev/binary/docker build --help'.
```
isn't nearly as useful as what we used to show, which was the full --help output.

**Comment by jasinner** [comment]
@duglin This change prints a short usage output, with the format of the command, when incorrect arguments are passed to a command. Like you said, currently at the moment the usage output is not very friendly as it only shows the number of argument required. This change not only shows that output, but the format of the command passed into the flags.subCmd method.

If users want the full, verbose output of the the flag.Usage command they can call --help.

I agree that the exit code on shortUsage should be 1. I will change it on Monday if required.

File: api/client/cli.go [hunk 1]
```diff
@@ -91,19 +91,26 @@ func (cli *DockerCli) Subcmd(name, signature, description string, exitOnError bo
...
-		}
 		fmt.Fprintf(cli.out, "\nUsage: docker %s%s%s\n\n%s\n\n", name, options, signature, description)
 		flags.SetOutput(cli.out)
 		flags.PrintDefaults()
 		os.Exit(0)
 	}
+	flags.ShortUsage = func() {
+		fmt.Fprintf(cli.out, "\nUsage: docker %s%s%s\n\n%s\n\n", name, options, signature, description)
+		flags.SetOutput(cli.out)
+		os.Exit(0)
```
Comment (SvenDowideit) [inline api/client/cli.go]: remove the Exit from here

File: utils/flags.go [hunk 2]
```diff
@@ -41,5 +42,4 @@ func ReportError(cmd *flag.FlagSet, str string, withHelp bool) {
 		}
 	}
 	fmt.Fprintf(cmd.Out(), "docker: %s.\n", str)
-	os.Exit(1)
```
Comment (SvenDowideit) [inline utils/flags.go]: put this one back

**Comment by SvenDowideit** [comment]
@jasinner yes please :)

**Comment by jasinner** [comment]
Hi guys, I added some more commits. Please review. 
When running 'docker run' without arguments output should be like this:
```
[jshepher@jshepher docker]$ bundles/1.5.0-dev/binary/docker-1.5.0-dev run 
docker: "run" requires a minimum of 1 argument. See 'bundles/1.5.0-dev/binary/docker-1.5.0-dev run --help'.

Usage: docker run IMAGE [COMMAND] [ARG...]

Run a command in a new container
```

File: api/client/cli.go [hunk 3]
```diff
@@ -91,19 +91,25 @@ func (cli *DockerCli) Subcmd(name, signature, description string, exitOnError bo
...
+	}
 	flags.Usage = func() {
-		options := ""
-		if signature != "" {
-			signature = " " + signature
-		}
-		if flags.FlagCountUndeprecated() > 0 {
-			options = " [OPTIONS]"
-		}
 		fmt.Fprintf(cli.out, "\nUsage: docker %s%s%s\n\n%s\n\n", name, options, signature, description)
```
Comment (duglin) [inline api/client/cli.go]: To ensure consistency, should these 2 lines be replaced with a call to `flags.ShortUsage()` ?

File: utils/flags.go [hunk 4]
```diff
@@ -1,45 +0,0 @@
-package utils
```
Comment (duglin) [inline utils/flags.go]: Why was this removed?  I'm guessing it was a mistake?

**Comment by duglin** [comment]
When done, please merge your commits down to one.

**Comment by duglin** [comment]
@jasinner ping

**Comment by jasinner** [comment]
@duglin The file utils/flags.go was removed from upstream master, so it was removed from my commit. 
I did as you suggested, and called ShortUsage from the Usage method.
I think I managed to squash my commits down to 1 now too.

**Comment by duglin** [comment]
@jasinner I think you lost your "signed-off by" stuff during your squashing.  Try doing:
```
git -s --amend --no-edit
```
and then pushing again to see if that fixes it.
Other than that I think it looks good

**Comment by jasinner** [comment]
@duglin added signed-off by

**Comment by duglin** [comment]
thanks!
LGTM
/cc @tiborvass @LK4D4

File: api/client/cli.go [hunk 5]
```diff
@@ -91,19 +91,24 @@ func (cli *DockerCli) Subcmd(name, signature, description string, exitOnError bo
 		errorHandling = flag.ContinueOnError
 	}
 	flags := flag.NewFlagSet(name, errorHandling)
+	options := ""
+	if signature != "" {
+		signature = " " + signature
+	}
+	if flags.FlagCountUndeprecated() > 0 {
+		options = " [OPTIONS]"
+	} else {
```
Comment (cpuguy83) [inline api/client/cli.go]: This is a redundant else.

File: api/client/cli.go [hunk 6]
```diff
@@ -91,19 +91,24 @@ func (cli *DockerCli) Subcmd(name, signature, description string, exitOnError bo
 		errorHandling = flag.ContinueOnError
 	}
 	flags := flag.NewFlagSet(name, errorHandling)
+	options := ""
+	if signature != "" {
+		signature = " " + signature
+	}
+	if flags.FlagCountUndeprecated() > 0 {
```
Comment (cpuguy83) [inline api/client/cli.go]: May help to move this up underneath the `options` declaration... or move `options := ""` just above this instead.

**Comment by cpuguy83** [comment]
As is here is the output from `docker save`:

```
docker: "save" requires a minimum of 1 argument. See 'docker save --help'.

Usage: docker save IMAGE [IMAGE...]

Save an image(s) to a tar archive (streamed to STDOUT by default)
```

And the output of `docker save --help` is litterally that same message minus the first line.
Can we get rid of the `See 'docker save --help'.` part, since it's the same exact output anyway?

**Comment by jasinner** [comment]
Passing --help to save is not the same output, it shows all the arguments. 
```
[jshepher@jshepher docker]$ bundles/1.5.0-dev/binary/docker-1.5.0-dev save
docker: "save" requires a minimum of 1 argument. See 'bundles/1.5.0-dev/binary/docker-1.5.0-dev save --help'.

Usage: docker save IMAGE [IMAGE...]

Save an image(s) to a tar archive (streamed to STDOUT by default)
```
with --help
```
[jshepher@jshepher docker]$ bundles/1.5.0-dev/binary/docker-1.5.0-dev save --help

Usage: docker save IMAGE [IMAGE...]

Save an image(s) to a tar archive (streamed to STDOUT by default)

  --help=false       Print usage
  -o, --output=      Write to an file, instead of STDOUT

```

**Comment by jasinner** [comment]
@cpuguy83 not actually sure what you meant by "move this up underneath the options declaration", but I moved the options 'if' check up under the options declaration. Also, I removed the redundant 'else'.

**Comment by cpuguy83** [comment]
Needs a gofmt.

**Comment by duglin** [comment]
@jasinner can you update the code by using gofmt per @cpuguy83's comment ?

**Comment by jfrazelle** [comment]
maybe someone should carry as there has been no comment in 7 days

**Comment by duglin** [comment]
will do if we don't get a response by later tonight

**Comment by jfrazelle** [comment]
cool thanks :)

**Comment by jasinner** [comment]
Hey guys,
Sorry for the delay. I'll try to run the format, and pull the latest
updates later today, Australia time.

**Comment by duglin** [comment]
What the threat of me actually coding will do!  :-)

**Comment by jasinner** [comment]
Ok, I think I fixed it up now.

**Comment by cpuguy83** [comment]
*sigh* looks like there's a test checking the output of `docker search --help`.

**Comment by duglin** [comment]
@jasinner I did some digging on this and you need to move the if-statement at https://github.com/docker/docker/pull/11858/files#diff-bcf4431eed51e55b4dca7deafdca48b0R115 into the Usage() func.  This is because you don't know what the various flags are until long after this func has exited, but you *will* know by the time the usage func is called.

**Comment by duglin** [comment]
@jasinner will you have time to fix this one up?

**Comment by jasinner** [comment]
Not until the week of 11th of May.

**Comment by duglin** [comment]
Would you like me to carry it for you or wait?

**Comment by jasinner** [comment]
Sorry for the delay in responding. I'd like to give it a go the week I'm
back. I'll get to you again after I return.

Jason

**Comment by duglin** [comment]
@jasinner okie dokie - thanks for letting us know

**Comment by jfrazelle** [comment]
any update :)

**Comment by jasinner** [comment]
@jfrazelle @duglin Hi guys, I moved the flag count check into the Usage funtion now, and the integration test seems to pass for me. 
Please check it, Thanks.

File: pkg/mflag/flag.go [hunk 7]
```diff
@@ -1086,7 +1095,6 @@ func (cmd *FlagSet) ReportError(str string, withHelp bool) {
 		}
 	}
 	fmt.Fprintf(cmd.Out(), "docker: %s.\n", str)
-	os.Exit(1)
```
Comment (duglin) [inline pkg/mflag/flag.go]: Why was this removed?  Where does the exit(1) happen now if not here?

Comment (duglin) [inline pkg/mflag/flag.go]: I see, you added an exit(1) above, but there are other calls to ReportError- like in CmdCreate - won't those break?

Comment (duglin) [inline pkg/mflag/flag.go]: Try a bogus cmd like:
```
docker create --net=ads --mac-address 123 ubuntu
```
and you'll see the issue.

**Comment by jasinner** [comment]
What if we move the call to os.Exit(0) back into the methods that call ReportError?

**Comment by duglin** [comment]
@jasinner can you run your code thru gofmt?  janky isn't happy about the current formatting.

As for moving the os.Exit() calls - I think that's the right solution.

**Comment by jasinner** [comment]
Hi @duglin and @GordonTheTurtle I run gofmt, and attempted to fix up the signature. I hope I got everything correct, but let me know if there is anything outstanding. Jason

**Comment by duglin** [comment]
@jasinner you'll need to squash your commits down to one. That should make Gordon happy, assuming that one commit is signed.

**Comment by thaJeztah** [comment]
ping @duglin I see this PR was updated, PTAL

**Comment by duglin** [comment]
@jasinner it looks good.  Sorry for not mentioning this earlier, but I think we need a few testcases added. In particular, I think ones that:
- test that the short usage is shown on a missing param (just short, not full help)
- test that we get the correct exit code in both short and normal help output

**Comment by jfrazelle** [comment]
@duglin do  you maybe want to help with the test cases so we can get this into 1.7 or @jasinner do you think you will be able to make these changes

**Comment by duglin** [comment]
@jfrazelle I'll work on the testcases either tonight or tomorrow.  if @jasinner doesn't respond by, say, Sunday maybe we can grab both sets of changes.

**Comment by duglin** [comment]
Or do you need it sooner?

**Comment by jfrazelle** [comment]
thanks SGTM

**Comment by duglin** [comment]
@jasinner I just did PR #13428 to continue this PR - mainly to add new testcases.  I did however find some oddities in the way some things were done for some commands, so I fixed those too.  Now we should be consistent.  Hope you're ok with this.  We're coming up on the deadline for v1.7 and we wanted to make sure this made it in.

**Comment by calavera** [comment]
This has been merged on master as part of #13428. Closing! :smile:

**Comment by thaJeztah** [comment]
@vieux @jfrazelle according to https://github.com/docker/docker/pull/11858#issuecomment-104803726 you wanted this in the 1.7 release?

**Comment by jfrazelle** [comment]
ill cherry pick it

**Comment by jasinner** [comment]
@duglin thanks for finishing this off. Happy to see it get merged.
